python/drned_xmnr/op/parse_log_errors.py

```diff
@@ -0,0 +1,71 @@
+def gather_problems(file):
+    problems = []
+    pending_pattern = None
+
+    matchless_max = None
+    matchless_to_go = None
+
+    for line_num, line in enumerate(file):
+        # TODO - beware - matcher with "lines-after" may interrupt
+        # other matcher if followed immediately in the log
+        matcher = _get_matcher(line)
+        if not matcher:
+            #  parse prefix-less lines for previously matched problem
+            if matchless_max and matchless_to_go:
+                message = _strip_problem_pattern(' - ', line)
+                if matchless_to_go == matchless_max:
+                    problems[-1].append((None, None, "---- details ----"))
+                # TODO - could use generic matcher for date on any line
+                detail_record = (line_num, None, message)
+                problems[-1].append(detail_record)
+                matchless_to_go -= 1
+                if matchless_to_go == 0:
+                    problems[-1].append((None, None, "---- more lines in log ----"))
+                continue
+
+            pending_pattern = None
+            matchless_max = None
+            matchless_to_go = None
+            continue
+
+        # handlers/data to control parsed line processing
+        (current_pattern, date_parser, matchless_max) = matcher
+        matchless_to_go = matchless_max
+
+        date = date_parser(line) if date_parser else None
+        message = _strip_problem_pattern(current_pattern, line)
+        problem_record = (line_num, date, message)
+
+        # continuation of pending problem
+        if current_pattern == pending_pattern:
+            problems[-1].append(problem_record)
+        # start of new problem
+        else:
+            problems.append([problem_record])
+            pending_pattern = current_pattern
+
+    return problems
+
+
+def _get_matcher(line):
+    # may want to extend to small class?
+    # Tuple of: (
+    #       - match pattern,
+    #       - date extractor fn,
+    #       - number of extra lines after last pattern match to include
+    # )
+    problem_matchers = [
+        (" !! ", lambda x: x.split()[1], 5),
+        ("E  ", None, None),
+        ("sync-from failed:", None, None),
```

Comment:
Not sure I understand: why don't you use the log date for the latter two matchers?

Reply:
fixed for `sync-from` matcher, leaving `E  ` matcher `None` due to multi-line & no-date-in-prefix format of the specific error in logs...